README.md
# Wordup for Visual Studio Code

The Wordup extension makes it easy to develop, manage and export WordPress plugins or themes from Visual Studio Code.

## Key Features

* Manage all your local WordPress projects directly in VSCode 
* Set up a new docker based development environment for your WordPress plugins/themes within minutes.
* Automatically install all necessary plugins and themes for WordPress. Even from local or remote zip-files.
* Export your plugin or theme to a zip-file. Or even export your whole project.

### Manage your WordPress projects:

![Manage your WordPress projects](resources/project-demo.gif)

### Create a new WordPress project:

![Create new WordPress plugin](resources/create-demo.gif)

> Tip: You can manage your WordPress installation settings always in the .wordup/config.yml of your project.

## Requirements

You will need to install [Docker](https://docs.docker.com/install/) and [wordup-cli](https://www.npmjs.com/package/wordup-cli)  on your machine and set up on the system path.

## For more information

* [Official website](https://wordup.dev): The official wordup website
* [Documentation](https://docs.wordup.dev): The wordup documentation
* [wordup-cli](https://github.com/wordup-dev/wordup-cli): The wordup CLI

## Release Notes

See the [change log](CHANGELOG.md).

## License

See the [LICENSE](LICENSE) file for details


CHANGELOG.md
# Change Log

All notable changes to the "wordup-code" extension will be documented in this file.


## [0.3.1][0.3.2] - 2019-07-19

- Bugfix: Wrong Windows filepath output
- Add installation as type for an new project

## [0.3.0] - 2019-06-27

- Support for new wordup config system
    -  wordup-cli >= 0.5.0 should be used
- wordup-cli update notification in vscode

## [0.2.0] - 2019-06-13

- Removed port restrictions, because the wordup-cli now assigns new ports automatically
    - INFO: There is no update notification, but users should have installed wordup-cli > 0.4.0
- Added Button to install dev server after init

## [0.1.0] - 2019-05-27

- Initial release

src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';

import { WordupProjectView } from './projectView';
import { WordupCli } from './wordupCli';

interface WordupTaskDefinition extends vscode.TaskDefinition {
	/**
	 * The task name
	 */
	task: string;

}

let taskProvider: vscode.Disposable | undefined;

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

	// Setup the project view
    const projectView = new WordupProjectView(context);
    
    // Setup wordup cli commands
	const wordupCli = new WordupCli(context, projectView);

    //Add wordup status bar
    const wordupStatusBarItem = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 100);
    wordupStatusBarItem.command = 'wordup.fastSelect';

	context.subscriptions.push(wordupStatusBarItem);

    // Update status bar item based on events for multi root folder changes
    if (vscode.workspace.workspaceFolders) {
        let wordupTaskPromise: Thenable<vscode.Task[]> | undefined = undefined;
        let pattern = path.join(vscode.workspace.workspaceFolders[0].uri.path,'.wordup', 'config.yml');
	    let fileWatcher = vscode.workspace.createFileSystemWatcher(pattern);
        fileWatcher.onDidChange(() => {
            updateStatusBarItem(wordupStatusBarItem);
            wordupTaskPromise = undefined;
        });
        fileWatcher.onDidCreate(() => {
            updateStatusBarItem(wordupStatusBarItem);
            wordupTaskPromise = undefined;
        });
        fileWatcher.onDidDelete(() => {
            updateStatusBarItem(wordupStatusBarItem);
            wordupTaskPromise = undefined;
        });
        taskProvider = vscode.tasks.registerTaskProvider('wordup', {
            provideTasks: () => {
                if (!wordupTaskPromise) {
                    wordupTaskPromise = getWordupTasks();
                }
                return wordupTaskPromise;
            },
            resolveTask(_task: vscode.Task): vscode.Task | undefined {
                return undefined;
            }
        });

    }
    
    // update status bar item once at start
    updateStatusBarItem(wordupStatusBarItem);
    
}

// this method is called when your extension is deactivated
export function deactivate() {
    if (taskProvider) {
		taskProvider.dispose();
	}
}

async function updateStatusBarItem(status: vscode.StatusBarItem) {
    const info = await isWordupProject();

	if (info) {
		status.text = `$(gear) Wordup`;
		status.show();
	} else {
		status.hide();
	}
}

async function isWordupProject(checkLocal?:boolean): Promise<boolean> {

    // If no workspace is opened or just a single folder, we return without any status label
    // because our extension only works when more than one folder is opened in a workspace.

    if (!vscode.workspace.workspaceFolders) {
        return Promise.resolve(false);
    }

    const resource = await vscode.workspace.findFiles('.wordup/config.yml');

    // If we have a file:// resource we resolve the WorkspaceFolder this file is from and update
    // the status accordingly.
    return new Promise((resolve, reject) => {
        resolve(resource.length > 0);
    });
}

async function getWordupTasks(): Promise<vscode.Task[]> {
	let workspaceRoot = vscode.workspace.workspaceFolders;
	let emptyTasks: vscode.Task[] = [];
	if (!workspaceRoot) {
		return emptyTasks;
	}
	if (!await isWordupProject()) {
		return emptyTasks;
    }

    let result: vscode.Task[] = [];
    let exportKind: WordupTaskDefinition = {
        label: 'Export the wordup src',
        type: 'wordup',
        task: 'export'
    };

    let exportTask = new vscode.Task(exportKind, vscode.TaskScope.Workspace, 'export', 'wordup', new vscode.ShellExecution('wordup export'));
    exportTask.group = vscode.TaskGroup.Build;
    exportTask.isBackground = false;
    result.push(exportTask);

    /*let installKind: WordupTaskDefinition = {
        label: 'Install the wordup components',
        type: 'wordup',
        task: 'install'
    };

    let installTask = new vscode.Task(installKind, 'install', 'wordup', new vscode.ShellExecution('npx wordup install'));
    result.push(installTask);
    */


    return result;

}


webpack.config.js
//@ts-check

'use strict';

const path = require('path');

/**@type {import('webpack').Configuration}*/
const config = {
  target: 'node', // vscode extensions run in a Node.js-context

  entry: './src/extension.ts', // the entry point of this extension
  output: {
    // the bundle is stored in the 'dist' folder (check package.json)
    path: path.resolve(__dirname, 'dist'),
    filename: 'extension.js',
    libraryTarget: 'commonjs2',
    devtoolModuleFilenameTemplate: '../[resource-path]'
  },
  devtool: 'source-map',
  externals: {
    vscode: 'commonjs vscode' // the vscode-module is created on-the-fly and must be excluded. Add other modules that cannot be webpack'ed, 📖 -> https://webpack.js.org/configuration/externals/
  },
  resolve: {
    // support reading TypeScript and JavaScript files
    extensions: ['.ts', '.js']
  },
  module: {
    rules: [
      {
        test: /\.ts$/,
        exclude: /node_modules/,
        use: [
          {
            loader: 'ts-loader'
          }
        ]
      }
    ]
  }
};
module.exports = config;

src/projectView.ts
import * as vscode from 'vscode';
import * as crypto from 'crypto';
import * as path from 'path';

import { WordupInitWebView } from './webview/wordupInit';
import { wordupConformPath } from './utils';

const shell = require('shelljs');
const Configstore = require('configstore');

interface ProjectRootData {
	id: string; 
	name: string; 
	slugName:string;
	path:string;
	installedOnPort:string;
	listeningOnPort:string;
}

interface ProjectInfoData {
	id: string; 
	name: string; 
	port: string;
	path:string;
	projectName:string;
}

interface ProjectNode {
	data: ProjectRootData | ProjectInfoData;
}

const fastSelectShowItem: vscode.QuickPickItem  = {label:'Show current project in explorer'};
const fastSelectOpenItem: vscode.QuickPickItem  = {label:'Open current project in browser'};
const fastSelectTasks: vscode.QuickPickItem  = {label:'Show wordup tasks (e.g. export)'};
const fastSelectStartItem: vscode.QuickPickItem  = {label:'Start new project'};


async function checkPackageUpdate(){
	
	const notifierConfig = new Configstore('update-notifier-wordup-cli');
	if(notifierConfig.size > 0){
		if(notifierConfig.has('update')){
			vscode.window.showInformationMessage('There is an update ('+notifierConfig.get('update.latest')+') available for [wordup-cli](https://www.npmjs.com/package/wordup-cli)', ...['Install']).then(selection => {
				if(selection === 'Install'){
					const terminal = vscode.window.createTerminal();
					terminal.sendText('npm install -g wordup-cli', false);
					terminal.show(true);
				}
			});
		}
	}
}

export function isRootData(object: any): object is ProjectRootData {
    return 'slugName' in object;
}

export class WordupProjectView {

	private projectExplorer: vscode.TreeView<ProjectNode>;
	public projects: ProjectNodeProvider;


	constructor(context: vscode.ExtensionContext) {


		this.projects = new ProjectNodeProvider(context);
		
		this.projectExplorer = vscode.window.createTreeView('wordupProjectView', { treeDataProvider: this.projects, showCollapseAll: false });

		this.projectExplorer.onDidChangeVisibility(e => {
        	//console.log( this.projectExplorer.visible);
    	});

		vscode.commands.registerCommand('wordup.fastSelect', async (selected?:vscode.QuickPickItem) => {

			if(!selected){
				selected = await vscode.window.showQuickPick([fastSelectShowItem,fastSelectOpenItem,fastSelectTasks,fastSelectStartItem],{ 
					canPickMany:false
				});
			}

			if(selected === fastSelectShowItem || selected === fastSelectOpenItem){

				const defaultPath = vscode.workspace.workspaceFolders ? vscode.workspace.workspaceFolders[0].uri.fsPath : '';
				if(defaultPath){
					const element = await this.findProjectByPath(defaultPath);
					
					if(element){
						if(selected === fastSelectShowItem){
							await this.projectExplorer.reveal(element, { focus: true, select: false, expand: true });
						}else if(isRootData(element.data)){
							if(element.data.listeningOnPort){
								vscode.env.openExternal(vscode.Uri.parse('http://localhost:'+element.data.listeningOnPort));
							}else {
								vscode.window.showInformationMessage('Project server is not running');
							}
						}
					}
				}
			}else if(selected === fastSelectTasks){
				vscode.commands.executeCommand('workbench.action.tasks.runTask');
			}else if(selected === fastSelectStartItem){
				vscode.commands.executeCommand('wordupProjectView.addEntry');
			}
		});


		vscode.commands.registerCommand('wordupProjectView.refreshEntry', () => this.projects.refresh());
		vscode.commands.registerCommand('wordupProjectView.editEntry', (node: any) => {
			if(node && isRootData(node.data)){
				vscode.commands.executeCommand('vscode.openFolder', vscode.Uri.file(node.data.path), true);
				vscode.window.showInformationMessage(`Opened wordup project: ${node.data.name}.`);
			}

		});

		vscode.commands.registerCommand('wordupProjectView.addEntry', async () => {
			const wordupInitForm = new WordupInitWebView(context);
		});


		vscode.commands.registerCommand('wordup.openInBrowser', async (node:any) =>  {
			if(node){
				vscode.env.openExternal(vscode.Uri.parse('http://localhost:'+node.data.port));
			}else{
				vscode.commands.executeCommand('wordup.fastSelect', fastSelectOpenItem);
			}
		});
		

	}


	public async findProjectByPath(aPath:string):Promise<ProjectNode | undefined>  {

		const id = crypto.createHash('sha1').update(wordupConformPath(aPath)).digest('hex');
		let allProjects = this.projects.cachedProjects;

		if(allProjects.length === 0){
			allProjects = await this.projects.getChildren();
		}

		const project = allProjects.find((element:ProjectNode) => {
			return (element.data.id === id);
		});

		return Promise.resolve(project);
	}

	public async runningProjects(port?:string):Promise<ProjectNode | undefined>  {
		let allProjects = this.projects.cachedProjects;

		if(allProjects.length === 0){
			allProjects = await this.projects.getChildren();
		}

		const runningProject = allProjects.find((element:ProjectNode) => {
			if(isRootData(element.data)){
				if(port){
					if(port === element.data.listeningOnPort){
						return true;
					}
				}else if(element.data.listeningOnPort){
					return true;
				}
			}
			return false;
		});
		return Promise.resolve(runningProject);
	}


}

export class ProjectNodeProvider implements vscode.TreeDataProvider<ProjectNode> {

	private _onDidChangeTreeData: vscode.EventEmitter<ProjectNode | undefined> = new vscode.EventEmitter<ProjectNode | undefined>();
	readonly onDidChangeTreeData: vscode.Event<ProjectNode | undefined> = this._onDidChangeTreeData.event;

	public cachedProjects:ProjectNode[] = [];
	readonly extensionPath:string;


	constructor(context: vscode.ExtensionContext) {
		this.extensionPath = context.extensionPath;
	}

	refresh(): void {
		this._onDidChangeTreeData.fire();
	}

	getTreeItem(element:ProjectNode): vscode.TreeItem {
		if(isRootData(element.data)){

			const wordupUri = vscode.Uri.file('/wordup/project#/'+element.data.id);

			const status = [];
			if(element.data.installedOnPort){
				status.push('Installed');
			}
			if(element.data.listeningOnPort){
				status.push('Running');
			}

			return {
				label:element.data.name,
				collapsibleState: vscode.TreeItemCollapsibleState.Collapsed,
				description: status.length > 0 ? '✔ '+status.join(' & ') : '',
				tooltip:'Slug: '+element.data.slugName,
				contextValue:'project',
				resourceUri:wordupUri,
				iconPath:{
					light: path.join(this.extensionPath, 'resources', 'light', 'project.svg'),
					dark: path.join(this.extensionPath, 'resources', 'dark', 'project.svg')
				}
			};
		}else{

			return {
				label: element.data.name,
				collapsibleState: vscode.TreeItemCollapsibleState.None,
				contextValue:element.data.id
			};

		}


	}


	getChildren(element?:ProjectNode):ProjectNode[] | Thenable<ProjectNode[]> {

		if (element) {
			if(isRootData(element.data)){
				return this.getProjectInfo(element.data);
			}
			return Promise.resolve([]);
		} else {
			return Promise.resolve(this.getProjects());
			
		}

	}

	getParent(element:ProjectNode) {
		return null;
	}

	private getProjects(): Promise<ProjectNode[]> {
		this.cachedProjects = [];

		return new Promise((resolve, reject) => {


			if(!shell.which('wordup')){
				vscode.window.showInformationMessage('We could not find the [wordup-cli](https://www.npmjs.com/package/wordup-cli) package on your system', ...['Install']).then(selection => {
					if(selection === 'Install'){
						const terminal = vscode.window.createTerminal();
						terminal.sendText('npm install -g wordup-cli',false);
        				terminal.show(true);
					}
				});
				resolve([]);
			}else{
				checkPackageUpdate();
			}

			shell.exec('wordup list --json --clear', {cwd:this.extensionPath}, (code:number, stdout: string, stderr: string) => {
				const wordup_list = JSON.parse(stdout);
				let plist:ProjectNode[] = [];
				wordup_list.forEach((element: ProjectRootData) => {
					//const newProject = new Project(element.name, element, vscode.TreeItemCollapsibleState.Collapsed);
					const newProject: ProjectNode = {data:element};
					this.cachedProjects.push(newProject);
					plist.push(newProject);
				});
				resolve(plist);
			});
			
		});
	}

	private getProjectInfo(element: ProjectRootData):ProjectNode[] {
		let name = 'Server: http://localhost:'+element.listeningOnPort;
		let context = 'projectRunning';

		if(!element.installedOnPort){
			name = 'Server: Not installed';
			context = 'projectStopped';
		}else if(!element.listeningOnPort){
			name = 'Server: Paused';
			context = 'projectPaused';
		}

		const info: ProjectInfoData = {
			id: context,
			name:name,
			port:element.listeningOnPort || element.installedOnPort,
			path:element.path,
			projectName:element.name
		};

		return [{data:info}];
	}


}

src/webview/wordupInit.ts

import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';
import * as cp from 'child_process';
import { getOutputChannel } from '../utils';
const slugify = require('slugify');

import { wordupConformPath } from '../utils';

interface ProjectInitFields {
	name: string; 
	path:string;
	type:string;
	title:string;
	admin:string;
    password:string;
    email:string;
    plugins:object;
}


export class WordupInitWebView {
    public panel:vscode.WebviewPanel;
	private readonly _extensionPath: string;

    constructor(context: vscode.ExtensionContext){
        this._extensionPath = context.extensionPath;

        this.panel = vscode.window.createWebviewPanel(
            'wordupInit', // Identifies the type of the webview. Used internally
            'New wordup project', // Title of the panel displayed to the user
            vscode.ViewColumn.One, // Editor column to show the new webview panel in.
            {
                enableScripts: true
            } // Webview options. More on these later.
        );

        this.getWebviewContent();

        this.panel.webview.onDidReceiveMessage(
            message => {
              switch (message.command) {
                case 'submitForm':
                  this.actionSubmitForm(message.value);
                  return;
                case 'selectFolder':
                  this.actionSelectFolder();
                  return;
              }
            },
            undefined,
            context.subscriptions
        );

    }

    actionSubmitForm(fields:ProjectInitFields){

        if (!fs.existsSync(fields.path)) {
			vscode.window.showErrorMessage('The selected path does not exist.');
            return;
		}

        vscode.window.withProgress({
            location: vscode.ProgressLocation.Notification,
            title: "Create new wordup project",
            cancellable: true
        }, (progress, token) => {

            progress.report({ increment: 5 });

            setTimeout(() => progress.report({ increment: 15 }), 1000);
            setTimeout(() => progress.report({ increment: 15 }), 2000);
            setTimeout(() => progress.report({ increment: 15 }), 4000);
            setTimeout(() => progress.report({ increment: 10 }), 8000);
            setTimeout(() => progress.report({ increment: 5, message: " 💤" }), 16000);

            return new Promise(resolve => {
                getOutputChannel().clear();
			    getOutputChannel().show(true);

                let env = Object.create( process.env );
                env.WORDUP_INIT_PATH = wordupConformPath(fields.path);
                env.WORDUP_INIT_NAME = fields.name;
                env.WORDUP_INIT_TYPE = fields.type;

                const wpInstall = {
                    values: {
                        title: fields.title,
                        adminUser:fields.admin, 
                        adminPassword:fields.password, 
                        adminEmail:fields.email,
                        plugins:fields.plugins || {},
                        themes:{}
                    }
                };

                env.WORDUP_INIT_WP_INSTALL = Buffer.from(JSON.stringify(wpInstall)).toString('base64');

                let cpCall = cp.exec('wordup init --no-autoinstall', {cwd:this._extensionPath, env:env});
                
                token.onCancellationRequested(() => {
                    cpCall.kill();
                });

                cpCall.stdout.on('data', (data:any) => {
                    getOutputChannel().append(data);
                });

                cpCall.stderr.on('data', (data:any) => {
                    getOutputChannel().append(data);
                });

                cpCall.on('close', (code:number) => {
                    if(code !== 0){
                        vscode.window.showErrorMessage("Oops, something didn't work. Please take a look at the output window for more details.");
                    }else {
                        this.panel.dispose();
                        vscode.commands.executeCommand('wordupProjectView.refreshEntry');
                        vscode.window.showInformationMessage('Successfully init new wordup project.',...['Open in new window', 'Install dev server']).then(selection => {
                            const projectPath = path.join( fields.path , slugify(fields.name, {lower: true, remove: /[*+~%\<>/;.(){}?,'"!:@#^|]/g}));
                            if(selection === 'Open in new window'){
                                vscode.commands.executeCommand('vscode.openFolder', vscode.Uri.file(projectPath), true);
                            }else if(selection === 'Install dev server'){
                                vscode.commands.executeCommand('wordup.installDevServer', undefined, projectPath);
                            }
                        });
                    }
                    resolve();
                });
            });
        });

    }

    actionSelectFolder(){
        vscode.window.showOpenDialog({canSelectFiles:false, canSelectFolders:true, canSelectMany:false}).then(fileUri => {
            if (fileUri && fileUri[0]) {
                this.panel.webview.postMessage({ command: 'selectedFolder', value: fileUri[0].fsPath  });
            }
        });

    }

    getWebviewContent() {
        const htmlPathOnDisk = path.join(this._extensionPath, 'resources','html','wordupInit.html');

        fs.readFile(htmlPathOnDisk, 'utf8',(err, contents) => {
            this.panel.webview.html = contents;
        });
    }

}

src/webview/wordupInit/setup.ts

import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';
import * as cp from 'child_process';
import { getOutputChannel } from '../../wordupCli';
const slugify = require('slugify');


interface ProjectInitFields {
	name: string; 
	path:string;
	type:string;
	title:string;
	admin:string;
    password:string;
    email:string;
    plugins:object;
}


export class WordupInitWebView {
    public panel:vscode.WebviewPanel;
	private readonly _extensionPath: string;

    constructor(context: vscode.ExtensionContext){
        this._extensionPath = context.extensionPath;

        this.panel = vscode.window.createWebviewPanel(
            'wordupInit', // Identifies the type of the webview. Used internally
            'New wordup project', // Title of the panel displayed to the user
            vscode.ViewColumn.One, // Editor column to show the new webview panel in.
            {
                enableScripts: true
            } // Webview options. More on these later.
        );

        this.getWebviewContent();

        this.panel.webview.onDidReceiveMessage(
            message => {
              switch (message.command) {
                case 'submitForm':
                  this.actionSubmitForm(message.value);
                  return;
                case 'selectFolder':
                  this.actionSelectFolder();
                  return;
              }
            },
            undefined,
            context.subscriptions
        );

    }

    actionSubmitForm(fields:ProjectInitFields){

        if (!fs.existsSync(fields.path)) {
			vscode.window.showErrorMessage('The selected path does not exist.');
            return;
		}

        vscode.window.withProgress({
            location: vscode.ProgressLocation.Notification,
            title: "Create new wordup project",
            cancellable: true
        }, (progress, token) => {

            progress.report({ increment: 5 });

            setTimeout(() => progress.report({ increment: 15 }), 1000);
            setTimeout(() => progress.report({ increment: 15 }), 2000);
            setTimeout(() => progress.report({ increment: 15 }), 4000);
            setTimeout(() => progress.report({ increment: 10 }), 8000);
            setTimeout(() => progress.report({ increment: 5, message: " 💤" }), 16000);

            return new Promise(resolve => {
                getOutputChannel().clear();
			    getOutputChannel().show(true);

                let env = Object.create( process.env );
                env.WORDUP_INIT_PATH = fields.path;
                env.WORDUP_INIT_NAME = fields.name;
                env.WORDUP_INIT_TYPE = fields.type;

                const wpInstall = {
                    values: {
                        title: fields.title,
                        adminUser:fields.admin, 
                        adminPassword:fields.password, 
                        adminEmail:fields.email,
                        plugins:fields.plugins || {},
                        themes:{}
                    }
                };

                env.WORDUP_INIT_WP_INSTALL = Buffer.from(JSON.stringify(wpInstall)).toString('base64');

                let cpCall = cp.exec('npx wordup init', {cwd:this._extensionPath, env:env});
                
                token.onCancellationRequested(() => {
                    cpCall.kill();
                });

                cpCall.stdout.on('data', (data:any) => {
                    getOutputChannel().append(data);
                });

                cpCall.stderr.on('data', (data:any) => {
                    getOutputChannel().append(data);
                });

                cpCall.on('close', (code:number) => {
                    if(code !== 0){
                        vscode.window.showErrorMessage("Oops, something didn't work. Please take a look at the output window for more details.");
                    }else {
                        this.panel.dispose();
                        vscode.window.showInformationMessage('Successfully installed new wordup project',...['Open in new window']).then(selection => {
                            if(selection === 'Open in new window'){
                                const projectPath = path.join( fields.path , slugify(fields.name, {lower: true}));
                                vscode.commands.executeCommand('vscode.openFolder', vscode.Uri.file(projectPath), true);
                            }
                        });
                        vscode.commands.executeCommand('wordupProjectView.refreshEntry');
                    }
                    resolve();
                });
            });
        });

    }

    actionSelectFolder(){
        vscode.window.showOpenDialog({canSelectFiles:false, canSelectFolders:true, canSelectMany:false}).then(fileUri => {
            if (fileUri && fileUri[0]) {
                this.panel.webview.postMessage({ command: 'selectedFolder', value: fileUri[0].fsPath  });
            }
        });

    }

    getWebviewContent() {
        const htmlPathOnDisk = vscode.Uri.file(
			path.join(this._extensionPath, 'src','webview','wordupInit', 'index.html')
		);

        fs.readFile(htmlPathOnDisk.path, 'utf8',(err, contents) => {
            this.panel.webview.html = contents;
        });
    }

}

src/wordupCli.ts
import * as vscode from 'vscode';
import * as cp from 'child_process';
import * as fs from 'fs';
import * as path from 'path';

const YAML = require('yaml');

import { WordupProjectView, isRootData } from './projectView';
import { getOutputChannel,wordupConformPath } from './utils';

export class WordupCli {

    public terminal:  vscode.Terminal | null;
    public projectView: WordupProjectView;
    readonly defaultPath: string;
    readonly extensionPath: string;

    constructor(context: vscode.ExtensionContext, projectView:WordupProjectView){

        this.projectView = projectView;

        this.terminal = null;

        this.defaultPath = vscode.workspace.workspaceFolders ? wordupConformPath(vscode.workspace.workspaceFolders[0].uri.fsPath) : '';
        this.extensionPath = context.extensionPath;

        vscode.commands.registerCommand('wordup.installDevServer', async (node:any, directPath?:string) =>  {
            let aPath = this.defaultPath;
            if(directPath){
                aPath = directPath;
            }else if(node){
                aPath = node.data.path;
            }

            if(aPath){

                //Check if wpInstall is set
                fs.readFile(path.join(aPath,'.wordup', 'config.yml'), async (err, data) => {  
                    if (!err) {
                        const pjson = YAML.parse(data.toString('utf8'));

                        if(pjson.hasOwnProperty('wpInstall')){

                            //Check if other projects are running on this port
                            let runningProjects = undefined;
                            if(pjson.hasOwnProperty('port')){
                                runningProjects = await projectView.runningProjects(pjson.port);
                            } 
                            
                            if(runningProjects !== undefined){
                                vscode.window.showWarningMessage('A different project is running ('+runningProjects.data.name+')');
                            }else{
                                const addMsg = node ?  node.data.projectName+': '  : '';
                                this.execWordupCli('install', aPath, addMsg+'Successfully installed server', '⏳ Please be patient ');
                            }
                        }else {
                            this.execVscodeTerminal('wordup install', aPath);
                        }
                    }else{
                        this.execVscodeTerminal('wordup install', aPath);
                    }
                });   
            }
        });
    
        vscode.commands.registerCommand('wordup.startDevServer', async (node:any) =>  {
            const nodePath = node ? node.data.path : this.defaultPath;

            const selectedProject = await projectView.findProjectByPath(nodePath);
            if(selectedProject && isRootData(selectedProject.data)){

                //Install project, instead of starting 
                if(!selectedProject.data.installedOnPort){
                    vscode.commands.executeCommand('wordup.installDevServer', node);
                    return;
                }
                
                this.execWordupCli('start',  nodePath, selectedProject.data.name+': Successfully started server');
            }
        });
    
        vscode.commands.registerCommand('wordup.stopDevServer', (node:any) =>  {
            const nodePath = node ? node.data.path : undefined;
            const addMsg = node ?  node.data.projectName+': '  : '';

            this.execWordupCli('stop', nodePath, addMsg+'Successfully stopped server.');
        });

        vscode.commands.registerCommand('wordup.deleteDevServer', (node:any) =>  {
            const nodePath = node ? node.data.path : undefined;
            const addMsg = node ?  node.data.projectName+': '  : '';
            vscode.window.showWarningMessage('If you delete the server, all data on this server will be deleted. Your project source code will not been affected.',{modal:true}, ...['Delete']).then(selection => {
                if(selection === 'Delete'){
                    this.execWordupCli('stop --delete', nodePath, addMsg+'Successfully deleted server.');
                }
            });
        });

    }

    public execWordupCli(cmd:string, path?:string, successMsg?:string, infoMsg?:string){
        const projectPath = path ? path : this.defaultPath;
        if(!projectPath){
            vscode.window.showInformationMessage('No wordup project found');
        }else{
            
            vscode.window.withProgress({
                location: vscode.ProgressLocation.Notification,
                title: "Executing wordup-cli: wordup "+cmd,
                cancellable: true
            }, (progress, token) => {

                progress.report({ increment: 5 });

                setTimeout(() => progress.report({ increment: 15 }), 1000);
                setTimeout(() => progress.report({ increment: 15 }), 2000);
                setTimeout(() => progress.report({ increment: 15 }), 4000);
                setTimeout(() => progress.report({ increment: 10 }), 8000);
                setTimeout(() => progress.report({ increment: 5, message: (infoMsg ? " "+infoMsg : " 💤") }), 16000);

                return new Promise(resolve => {
                    getOutputChannel().clear();

                    let env = Object.create( process.env );
                    env.WORDUP_PROJECT_PATH = projectPath;

                    let cpCall = cp.exec('wordup '+cmd, {cwd:this.extensionPath, env:env});
                    
                    token.onCancellationRequested(() => {
                        cpCall.kill();
                    });

                    cpCall.stdout.on('data', (data:any) => {
                        getOutputChannel().append(data);
                    });

                    cpCall.stderr.on('data', (data:any) => {
                        getOutputChannel().append(data);
			            getOutputChannel().show(true);
                    });

                    cpCall.on('close', (code:number) => {
                        if(code !== 0){
                            vscode.window.showWarningMessage('We could not execute this command successfully. You can try it in your terminal. ', ...['Try in terminal']).then(selection => {
                                if(selection === 'Try in terminal'){
                                    this.execVscodeTerminal('wordup '+cmd, projectPath);
                                }
                            });
                        }else if(successMsg){
                            vscode.window.showInformationMessage(successMsg);
                            vscode.commands.executeCommand('wordupProjectView.refreshEntry');
                        }
                        resolve();
                    });
                });
            });

        }



    }


    public execVscodeTerminal(cmd:string, path:string){
        if(this.terminal){
            this.terminal.dispose();
        }

        this.terminal = vscode.window.createTerminal({name:'Wordup', cwd:path});
        this.terminal.sendText(cmd);
        this.terminal.show();
    }

}